### PR TITLE
Bulk publish logic

### PR DESCRIPTION
**Pull request by vargaat**
refs: MBL-17300
affects: Teacher
release note: none

test plan:
- Turn on feature flag.
- Start a bulk publish.
- Check if progress updates on the progress screen.
- Wait for it to complete.
- Go to modules screen, check if publish states are correctly updated or not.
- Repeat the publish but this time dismiss the progress screen and wait on the modules list.
- Progress indicators should disappear and published states should update.
- Repeat the publish but cancel the operation.

## Screenshots

<table>
<tr><th>Before</th><th>After</th></tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
</tr>
</table>

## Checklist

- [ ] Follow-up e2e test ticket created
- [ ] A11y checked
- [x] Tested on phone
- [ ] Tested on tablet
- [x] Tested in dark mode
- [ ] Tested in light mode
- [ ] Approve from product

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request will not generate a release note.</td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Teacher

[MBL-17300](https://instructure.atlassian.net/browse/MBL-17300)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 91.86% | 91.82% | 0.03%
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a4361c1f832cda9955220d540b6d22f285ac639a
</p>

**Comment by inst-danger** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

File: Core/Core/Modules/ModulesAssembly.swift [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        let interactor = ModulePublishInteractorLive(courseId: courseId)
+        publishInteractorsByCourseIds[courseId] = interactor
+
+        interactor
+            .statusUpdates
+            .receive(on: RunLoop.main)
+            .sink { update in
+                AppEnvironment.shared.topViewController?.findSnackBarViewModel()?.showSnack(update)
+            }
+            .store(in: &subscriptions)
```
Comment (szabinst) [inline Core/Core/Modules/ModulesAssembly.swift:39]: We might need to tinker about a best practice about where to place this kind of logic. Assembly seems a bit out of place for a UI related update. 

Maybe put it in the first view model that where the flow starts?

**Review by szabinst — APPROVED** [review]
QA + 1, good job on this feature! 

I had one observation, but it's not a blocker.

**Review by ndrsszsz — DISMISSED** [review]
QA +1

**Review by szabinst — APPROVED** [review]
QA + 1

**Review by ndrsszsz — APPROVED** [review]

**Review by rh12 — APPROVED** [review]
QA + 1
I found some minor things, but IMHO these could be improved later:

Navigating to a next page (Module Item) while some publish in progress and navigating back freezes the spinners. They return to published state icon once progress completes.

Unpublishing a Module+Items and quickly Publishing All Modules+Items:
1. Single module unpublish finishes
2. Single module + items show unpublished icons
3. AllModules publish still in progress
4. Single module + items are not showing spinners
5. Single module + items show published icons once AllModules progress finishes

**Comment by vargaat** [comment]
Nice finds!

> Navigating to a next page (Module Item) while some publish in progress and navigating back freezes the spinners. They return to published state icon once progress completes.

Yes, this is a bug in the spinner component itself and not related directly to this PR but still worth a fix.

> Unpublishing a Module+Items and quickly Publishing All Modules+Items:
> 
> 1. Single module unpublish finishes
> 2. Single module + items show unpublished icons
> 3. AllModules publish still in progress
> 4. Single module + items are not showing spinners
> 5. Single module + items show published icons once AllModules progress finishes

This is a known edge case that would have made the business logic around tracking a little more complicated so I just left it as-is. We can come back to this is this becomes a common usecase.

**Pull request merged**